segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,339 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+                    }, this.storageExecutor);
+        }
+        return CompletableFuture.completedFuture(null);
+    }
 
-        if (queueSize.get() >= config.getGarbageCollectionMaxQueueSize()) {
-            log.warn("{}: deleteGarbage - Queue full. Could not delete garbage. Chunks skipped", traceObjectId);
+    private CompletableFuture<Void> failTask(TaskInfo infoToRetire) {
+        if (null != taskQueue) {
+            return taskQueue.addTask(failedQueueName, infoToRetire);
         }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    private CompletableFuture<Void> deleteSegment(TaskInfo taskInfo) {
+        val streamSegmentName = taskInfo.getName();
+        ArrayList<String> chunksToDelete = new ArrayList<>();
+        @Cleanup
+        val txn = metadataStore.beginTransaction(false, streamSegmentName);
+        return txn.get(streamSegmentName)
+                .thenComposeAsync(storageMetadata -> {
+                    val segmentMetadata = (SegmentMetadata) storageMetadata;
+                    val failed = new AtomicReference<Throwable>();
+                    if (null == segmentMetadata) {
+                        log.debug("{}: deleteGarbage - Segment metadata does not exist. segment={}.", traceObjectId, streamSegmentName);
+                        return CompletableFuture.completedFuture(null);
+                    } else if (segmentMetadata.isActive()) {
+                        log.debug("{}: deleteGarbage - Segment is not marked as deleted. segment={}.", traceObjectId, streamSegmentName);
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        return new ChunkIterator(storageExecutor, txn, segmentMetadata)
+                                .forEach((metadata, name) -> {
+                                    metadata.setActive(false);
+                                    txn.update(metadata);
+                                    chunksToDelete.add(name);
+                                })
+                                .thenRunAsync(() -> this.addChunksToGarbage(txn.getVersion(), chunksToDelete), storageExecutor)
+                                .thenRunAsync(() -> deleteBlockIndexEntriesForChunk(txn, streamSegmentName, segmentMetadata.getStartOffset(), segmentMetadata.getLength()), storageExecutor)
+                                .thenComposeAsync(v -> {
+                                    txn.delete(segmentMetadata.getName());
+                                    return txn.commit();
+                                }, storageExecutor)
+                                .handleAsync((v, e) -> {
+                                    if (null != e) {
+                                        log.error(String.format("%s deleteGarbage - Could not delete metadata for garbage segment=%s.",
+                                                traceObjectId, streamSegmentName), e);
+                                        failed.set(e);
```

Comment:
Why do you need this variable? Could the `thenComposeAsync` just get as input whether the computation has completed successfully or not (e.g., boolean) and use it instead of having this `AtomicReference<Throwable>`? You are not event using the throwable object inside that variable in `thenComposeAsync`, right?

Reply:
Original intention was to treat transient errors differently. However it will be a bigger change so postponing that change and reverting back to using Boolean as before.